mantidproject/mantid
# quantify code coverage

This Fixes [#11832](http://trac.mantidproject.org/mantid/ticket/11832)

Measure C++ code coverage in Mantid using gcov. To filter out the generated CxxTest files, I use cmake to generate a whitelist of files to check and then submit the resulting json file to coveralls.io.

I am using portions of https://github.com/JoakimSoderberg/coveralls-cmake but couldn't use it entirely because `CoverallsGenerateGcov.cmake` was unable to correctly parse many Mantid source files. Ultimately, I ended up rewriting that file in python which appears to be more robust at parsing the resulting gcov files and includes a proper json encoder.

At the moment coverage is limited to framework, though this can be expanded at a later date.

**testing**:
- set the environment variable `COVERALLS_REPO_TOKEN` to something. The actual value will be set in Jenkins and shouldn't be in the source tree. 
- Starting with a clean build directory, run cmake with `CMAKE_BUILD_TYPE=Debug` and `COVERALLS=ON`. I recommend using make, as ninja hides the ctest output.
- run `make && make AllTests`
- run `make coveralls`. This step uses a lot of IO (writing 100-200 MB/sec) and is VERY SLOW.
- examine coveralls.json and verify the format matches the [coveralls api](https://coveralls.zendesk.com/hc/en-us/articles/201350799-API-Reference). The top-level arguments `git`, `repo_token` and `source_files` should be present. The argument `source_files` should be a very long list following the API's format.
